perf(app,robot-server): Download analyses as raw JSON documents

SyntaxColoring opened the pull request:
# Overview

This dramatically improves the load time for certain views in the Opentrons App and Flex on-device display.

For example, under the circumstances described in RSS-328, it reduces load time of the Flex's home page from ~5 minutes to <5 seconds.

This closes RSS-328 and RSS-160. It probably closes some other performance-related tickets. It also helps a little bit with RSS-98. 

# Architecture

Protocol analysis on the robot is a major known compute bottleneck for us. One of the problems is the overhead of storing the completed analyses in the database and then extracting them. It goes through several layers of translation:

```mermaid
flowchart LR
    subgraph robot-server
        analysis_engine((Analysis engine))
        subgraph SQL database
            bytes
        end
        dict
        pydantic[Pydantic object]
        fastapi((FastAPI endpoint))
        json_str
    end
    analysis_engine --> pydantic
    bytes <-->|pickle| dict
    dict <--> pydantic
    pydantic --> fastapi --> json_str[JSON string] --> to_client((HTTP client))
```

This scheme adds a *lot* of overhead. Creating Pydantic objects is especially slow. (This scheme is also very brittle—see RSS-98.)

To improve this, this PR adds a second path that's basically a direct read from the database:

```mermaid
flowchart LR
    subgraph robot-server
        analysis_engine((Analysis engine))
        subgraph SQL database
            bytes
            json_str_db[JSON string]
        end
        dict
        pydantic[Pydantic object]
        old_fastapi((FastAPI endpoint))
        new_fastapi((FastAPI endpoint))
        json_str
    end
    analysis_engine --> pydantic
    bytes <-->|pickle| dict
    dict <--> pydantic
    pydantic --> old_fastapi --> json_str[JSON string] --> to_client((HTTP client))
    json_str_db <--- pydantic
    json_str_db --> new_fastapi --> to_client
    linkStyle 7 stroke:red
    linkStyle 8 stroke:red
```

We expose this through an experimental new HTTP endpoint, `GET /protocols/:id/analyses/:id/asDocument`. We leave the existing endpoints, such as `GET /protocols/:id/analyses/:id`, untouched for now.

This requires a database migration to add a new column for the JSON string, so we do that.

# Detailed changelog

## Server

* Add a new endpoint. Document it as experimental.
* Add a new column to the `analysis` table: `completed_analysis_as_document`, which stores the serialized JSON as a `VARCHAR`.
* Introduce a new database migration to support the new column.
    * The migration includes copying over all existing data, so the new column should never be `NULL`.
    * Per previous discussion in RSS-130, we report an in-progress migration with an HTTP 503 status code on `GET /health` and on the endpoints that need the database.
    * Expected migration times:
        * Post-release Flexes: None, if we get this in for v7.0.0.
        * Internal Flexes: 5 minutes, from testing with the data in RSS-328.
        * OT-2s: 1 minute on average (from my old notes in RSS-130), or 10 minutes if heavily-loaded (extrapolating from RSS-328). There's some room for optimizing this.

## Desktop app and ODD

@b-cooper can provide more details, but roughly, we're switching a bunch of stuff to query the new fast endpoint and avoid querying the old slow endpoints.

Helpers for the new `GET /protocols/:protocolId/analyses/:analysisId/asDocument` endpoint have been added `api-client` and `react-api-client`.  The new `useProtocolAnalysisAsDocumentQuery` has been substituted across all locations in the Desktop and ODD where we were formerly requesting `GET /protocols/:protocolId/analyses` (old non-performant un-pickling path) and picking the latest entry.
 
# Test Plan

@SyntaxColoring tested on an OT-2 and Flex:

* [x] It should be faster. :) See RSS-328 for one test case.
* [x] A Flex's on-device display should say "initializing..." while the migration is in progress.
* [x] Per prior discussion in RSS-130, the desktop app may say something like "This robot's API server is not responding correctly to requests..." while the migration is in progress. This is potentially confusing, but it's what it's historically shown while the robot boots up. The only difference now is that bootup is taking longer.
* [x] Power-cycle a robot in the middle of the migration and make sure it doesn't cause any errors. The migration should start from scratch.
* [x] You should be able to freely upgrade and downgrade across this commit without anything breaking.

# Review requests

Architecturally, is there anything you think we need to do now in order to make this less risky, or avoid headaches in the long run?

# Risk assessment

Medium.

* Supporting both strategies simultaneously adds complexity to the server, which will be bug-prone if we let it linger.

* Any kind of robot-server database migration with our homegrown system is inherently kind of scary, and this is the first real one that we're doing in production.
  
  We have good automated tests for it (e.g. [test_tables.py](https://github.com/Opentrons/opentrons/blob/663f6d83fb13e2fc4411a98648731bc5f5f1bb46/robot-server/tests/persistence/test_tables.py) and [test_persistence.py](https://github.com/Opentrons/opentrons/blob/663f6d83fb13e2fc4411a98648731bc5f5f1bb46/robot-server/tests/integration/http_api/persistence/test_compatibility.py)), but things can still sneak in. Examples:

    * The transaction bug fixed in #13424
    * [This thing about a mismatched column type](https://github.com/Opentrons/opentrons/blob/663f6d83fb13e2fc4411a98648731bc5f5f1bb46/robot-server/robot_server/persistence/_migrations.py#L156-L160)
    * [This quirk in how we stamp the database with schema versions](https://opentrons.slack.com/archives/CSPMY9048/p1693446038126749)

SyntaxColoring committed:
Hack up some performance tests.

SyntaxColoring committed:
Add a migration to add the new column.

SyntaxColoring committed:
Try to migrate records eagerly.

b-cooper committed:
temporarily remove tracking from run card, and check equipment from analysis as doc

SyntaxColoring committed:
Delete benchmarking script.

b-cooper committed:
server initializing empty state for recent run protocol card

b-cooper committed:
remove old data fetching comments from required hardware hook

SyntaxColoring committed:
Merge branch 'chore_release-7.0.0' into performance_testing_analyses_as_opaque_documents

Resolve conflicts in:
* robot-server/robot_server/protocols/completed_analysis_store.py

b-cooper committed:
refactor useMostRecentCompletedAnalysis to use asDocument endpoint

b-cooper committed:
back off the top level maintenance run poll a bit

SyntaxColoring committed:
Add a Tavern test for the analysis endpoints.

SyntaxColoring committed:
Tweak serialization configurables to match what we normally do through FastAPI.

b-cooper commented on app/src/organisms/TakeoverModal/MaintenanceRunTakeover.tsx:
```diff
@@ -6,6 +6,7 @@ import {
 import { TakeoverModal } from './TakeoverModal'
 import { TakeoverModalContext } from './TakeoverModalContext'
 
+const MAINTENANCE_RUN_POLL_MS = 10000
```
Backing off this refetch interval to a slower poll.  There are separate interval instantiated within wizards where we need more up to date info

SyntaxColoring committed:
Add endpoint docs and fix Content-Type header.

b-cooper committed:
update all instances of analyses query to as doc

b-cooper committed:
undo problematic useprotocoldetailsforrun test fix

b-cooper committed:
fix up test for use protocol details for run

SyntaxColoring committed:
Unrelated test fixups.

SyntaxColoring committed:
Fix edge case with potentially NULL documents.

SyntaxColoring committed:
Add CompletedAnalysisStore unit tests.

b-cooper committed:
fix up formatting and tests

SyntaxColoring committed:
Import fixup.

SyntaxColoring committed:
Add checks for new endpoint in persistence snapshot compatibility test.

SyntaxColoring committed:
Add checks for new endpoint in persistence test.

SyntaxColoring committed:
Describe confusing schema stamp behavior.

SyntaxColoring committed:
Simpler solution for the upgrade-downgrade-upgrade edge case.

Separate the schema migration from the data migration. Do the data part unconditionally.

SyntaxColoring committed:
Document column data types.

SyntaxColoring committed:
Format 'n lint.

SyntaxColoring committed:
Update test_tables.py.

SyntaxColoring committed:
Delete test_migrations.py. :o

SyntaxColoring commented on robot-server/tests/persistence/test_migrations.py:
I'm deleting this file because these tests were difficult to update for new database schemas, and I don't think they were helpful.

These were old tests, added in #10309 when the database was quite new. Since then, we've added:

* Integration tests with snapshots of old persistence directories. PR #11448
* Safeties against accidentally changing SQL schema. PR #11536

Combined, I think those newer approaches supersede this old one.

SyntaxColoring commented on robot-server/robot_server/protocols/analysis_store.py:
```diff
@@ -193,6 +193,26 @@ async def get(self, analysis_id: str) -> ProtocolAnalysis:
+        )
+
+        if pending_analysis is not None:
+            # TODO: Probably want this to be a separate HTTP status code so the pending case
+            # doesn't have to pollute the completed analysis schema.
```
Reminder to self to resolve this.

SyntaxColoring committed:
Note potential future trap with _systemd_notify().

SyntaxColoring committed:
Raise 404 if the given analysis is still pending.

Add unit tests for AnalysisStore.

SyntaxColoring committed:
Minor fixups to comments and formatting.

SyntaxColoring committed:
we are all children of the machine. the machine takes care of us.

the machine understands `make format-js` as an expression of love.

vegano1 commented:
Sweet!

SyntaxColoring committed:
Bump copmatibility tests' startup timeouts, for CI.

SyntaxColoring committed:
Add missing router tests.

jbleon95 commented on robot-server/tests/protocols/test_protocols_router.py:
```diff
@@ -663,3 +664,70 @@ async def test_get_protocol_analysis_by_id_analysis_not_found(
+    protocol_store: ProtocolStore,
+    analysis_store: AnalysisStore,
+) -> None:
+    """It should get a single full analysis by ID."""
```
```suggestion
    """It should 404 if the analysis does not exist."""
```

SyntaxColoring replied:
🤦 Oops. Thanks for your attention to detail.

jbleon95 reviewed (approved):
Reviewed part of PR in live review, Python side looks good to me.

SyntaxColoring committed:
Fix test docstrings.

SyntaxColoring committed:
It would help if I applied the increased timeout to the correct file.

This reverts commit 9ebe7379b0542f4e4ee08df0af7739a429a05900.

smb2268 reviewed (approved):
All of the JS code changes look sound to me. I put this branch on app&ui bot and everything loaded very quickly!
